### ISSUE TITLE
pair-setup-pin failure

### ISSUE THREAD

**Issue by corecoding** [philippe44/libraop#13]
First of all, thank you for your work on the crypto authentication. I found your original perl demo over at https://htmlpreview.github.io/?https://github.com/philippe44/RAOP-Player/blob/master/doc/auth_protocol.html and almost got it working. My AppleTV brings up the pin code screen, but when I enter it, it says FAILED twice (first one being after the pair-setup-pin.

I tracked the code back to this repo, and saw that you had a pairing folder with the same perl code, but an independent lib folder. This one is not working for me at all yet, saying:
Math::BigInt: couldn't load specified math lib(s), fallback to Math::BigInt::Calc at lib/Crypt/SRP.pm line 10.

Perhaps it is because I am running Perl 5.26 and the lib folder only has static libraries up to 5.24?

Any help would be appreciated. I just need a working example for an open source Gnome Extension that I am working on.

**Comment by philippe44** [comment]
yes, you need to rebuild for your combination of OS/Perl version. You need Crypt::CryptX and Crypt::Ed25519

**Comment by corecoding** [comment]
I used cpan to install the required libraries. The script does cause the apple tv to present a pin code. However, the pin code is not accepted and the perl script says FAILED twice... Are your versions of the libraries modified from the system defaults?

![Screenshot from 2019-05-30 19-47-21](https://user-images.githubusercontent.com/3803591/58678231-cf79e400-8313-11e9-9d8b-69a626fab5e2.png)

**Comment by corecoding** [comment]
Got a little closer tonight... Comparing the values when $force=1, I was having a discrepancy with M1... Changing
my $client = Crypt::SRP->new('RFC5054-2048bit', 'SHA1');
To
my $client = Crypt::SRP->new({ group => 'RFC5054-2048bit', hash => 'SHA1', format => 'raw', appletv => 1 });
Gets rid of the M1 difference. Still not working... Getting closer though.

**Comment by philippe44** [comment]
Well there was a modification that I had to do but I submitted a change req to the SRP developper and he integrated it. That was a while ago. You can check the code/version I’m using in my LMS-to-raop bridge

**Comment by corecoding** [comment]
Would you mind doing me a favor? Can you test the pairing perl script to see if it still authenticates with your hardware? Before I spend the weekend trying to figure out why I am receiving a 500 error on pair-setup-pin . I know it is probably frustrating for me to ask, but it would be much appreciated.

**Comment by philippe44** [comment]
I tried on my old Win32 5.16 perl version and it works. I have
CryptX: 0.060 (CPAN version)
SRP: 0.017 (from my lib dir)
I know this can be very frustrating (I've spent a lot of time on that) 
Good luck!

**Comment by corecoding** [comment]
Very frustrating indeed. Still can't get it to work. I tried your modified SRP 0.017 and the new one 0.018 with the appletv => flag. I'm on the fence as to whether I want to move forward... The weird thing is that when $force is set to 1, the output is identical.. I think the difference may be another library, because aside from using your modified SRP 0.07, I am using CPAN installed libraries. My CryptX is 0.063.

[airplay.txt](https://github.com/philippe44/RAOP-Player/files/3244332/airplay.txt)

**Comment by philippe44** [comment]
Maybe you can send one example of calculation from your ATV and I'll check if I have the same result with $force=1 but your inputs. Did you dump what is the returned HTTP data in step 3 because I don't know what is the kind of failure. The other option is that you have different parameters setup in your ATV

**Comment by corecoding** [comment]
My Apple TV AirPlay is setup to allow anyone on the network to connect (no restrictions). Here are 3 detailed outputs:
[airplay.force.txt](https://github.com/philippe44/RAOP-Player/files/3244354/airplay.force.txt)
[airplay.mine.clean.txt](https://github.com/philippe44/RAOP-Player/files/3244355/airplay.mine.clean.txt)
[airplay.mine.txt](https://github.com/philippe44/RAOP-Player/files/3244356/airplay.mine.txt)

**Comment by philippe44** [comment]
arrghh ... did you try my script or did you always started from the doc? There is a typo in the doc. At step 2, when you send M1,  it's not a POST pair-pin-start, but a POST pair-setup-pin

[edit]: I guess that's a silly question as the log you're sending obviously is my perl script

**Comment by corecoding** [comment]
It is indeed your script. I did refactor it, but I promise it was not working even before I refactored it. As far as pair-pin-start and pair-setup-pin, my step 2/M1 is currently using pair-setup-pin.

**Comment by philippe44** [comment]
BTW, I'm using an old Perl 5.16 on Windows for my test

**Comment by corecoding** [comment]
Good news. I got it working! As I mentioned, I couldn't get your lib folder to work on my machine. It is an Ubuntu 18.04 box. During my testing/tweaking, one of the pair-setup-pin requests changed to SUCCESS. I thought I had changed something to make it work, but I was wrong. It worked one time, all by itself. This lead me to believe it was an encoding issue of some sort. I tracked it down to the BinaryWriter/BinaryReader. Something changed in the latest CPAN version, as when I swapped in your published version, everything works great. Even the new SRP library with appletv flag.

What a trip... But it's working! Thank you for your time.
